compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64ZeroMemoryOp.java

```diff
@@ -0,0 +1,182 @@
+        Register size = asRegister(lengthValue);
+        if (useDcZva && zvaLength > 0) {
+            // From ARMv8-A architecture reference manual D12.2.35 Data Cache Zero ID register:
+            // A valid ZVA length should be a power-of-2 value in [4, 2048]
+            assert (CodeUtil.isPowerOf2(zvaLength) && 4 <= zvaLength && zvaLength <= 2048);
+            emitZeroMemoryWithDc(masm, base, size, zvaLength);
+        } else {
+            // Use store pair instructions (STP) to zero memory as a fallback.
+            emitZeroMemoryWithStp(masm, base, size);
+        }
+    }
+
+    /**
+     * Zero a chunk of memory with DC ZVA instructions.
+     *
+     * @param masm the AArch64 macro assembler.
+     * @param base base an 8-byte aligned address of the memory chunk to be zeroed.
+     * @param size size of the memory chunk to be zeroed, in bytes, must be multiple of 8.
+     * @param zvaLength the ZVA length info of current AArch64 CPU.
+     */
+    private static void emitZeroMemoryWithDc(AArch64MacroAssembler masm, Register base, Register size, int zvaLength) {
+        Label preLoop = new Label();
+        Label zvaLoop = new Label();
+        Label postLoop = new Label();
+        Label tail = new Label();
+        Label done = new Label();
+
+        try (AArch64MacroAssembler.ScratchRegister sc1 = masm.getScratchRegister()) {
+            Register rscratch1 = sc1.getRegister();
+
+            // Count number of bytes to be pre-zeroed to align base address with ZVA length.
+            masm.neg(64, rscratch1, base);
+            masm.and(64, rscratch1, rscratch1, zvaLength - 1);
+
+            // Is size less than number of bytes to be pre-zeroed? Jump to POST_LOOP if so.
+            masm.cmp(64, size, rscratch1);
+            masm.branchConditionally(AArch64Assembler.ConditionFlag.LE, postLoop);
+            masm.sub(64, size, size, rscratch1);
+
+            // Pre-ZVA loop.
+            masm.bind(preLoop);
+            masm.subs(64, rscratch1, rscratch1, 8);
+            masm.branchConditionally(AArch64Assembler.ConditionFlag.LT, zvaLoop);
+            masm.str(64, zr, AArch64Address.createPostIndexedImmediateAddress(base, 8));
+            masm.jmp(preLoop);
+
+            // ZVA loop.
+            masm.bind(zvaLoop);
+            masm.subs(64, size, size, zvaLength);
+            masm.branchConditionally(AArch64Assembler.ConditionFlag.LT, tail);
+            masm.dc(AArch64Assembler.DataCacheOperationType.ZVA, base);
+            masm.add(64, base, base, zvaLength);
+            masm.jmp(zvaLoop);
+
+            // Handle bytes after ZVA loop.
+            masm.bind(tail);
+            masm.add(64, size, size, zvaLength);
+
+            // Post-ZVA loop.
+            masm.bind(postLoop);
+            masm.subs(64, size, size, 8);
+            masm.branchConditionally(AArch64Assembler.ConditionFlag.LT, done);
+            masm.str(64, zr, AArch64Address.createPostIndexedImmediateAddress(base, 8));
```

Comment:
Hi @pli-arm does ARM support unaligned str/stp? I am making changes to lift the alignment restriction here, and was wondering whether we need to align it first (e.g., when `base` is `0x*07`)?

Reply:
Yes, the architecture does support. But it's better to avoid unaligned memory access because it hurts the performance a lot (on some AArch64 CPU implementations).